src/ai/backend/manager/cleanup/session.py

```diff
@@ -0,0 +1,132 @@
+import asyncio
+import functools
+import logging
+from contextlib import asynccontextmanager as actxmgr
+from contextlib import suppress
+from datetime import datetime, timedelta
+from typing import Any, AsyncIterator, Mapping, Optional, Protocol, Sequence
+
+import aiotools
+import sqlalchemy as sa
+from dateutil.relativedelta import relativedelta
+from dateutil.tz import tzutc
+from sqlalchemy.orm import load_only, noload
+
+from ai.backend.common.events import KernelLifecycleEventReason
+from ai.backend.common.validators import TimeDelta
+from ai.backend.logging import BraceStyleAdapter
+
+from ..api.context import RootContext
+from ..config import session_hang_tolerance_iv
+from ..models import SessionRow, UserRole
+from ..models.session import SessionStatus
+from ..models.utils import ExtendedAsyncSAEngine
+
+__all__ = ("stale_session_collection_ctx",)
+
+log = BraceStyleAdapter(logging.getLogger(__spec__.name))
+
+
+class SessionDestroyer(Protocol):
+    async def destroy_session(
+        self,
+        session: SessionRow,
+        *,
+        forced: bool = False,
+        reason: Optional[KernelLifecycleEventReason] = None,
+        user_role: UserRole | None = None,
+    ) -> Mapping[str, Any]: ...
+
+
+def get_interval(
+    threshold: TimeDelta,
+    *,
+    max_interval: float = timedelta(hours=1).total_seconds(),
+    heuristic_interval_weight: float = 0.4,  # NOTE: to repeat more than twice within the same time window.
+) -> float:
+    if isinstance(threshold, relativedelta):  # months, years
+        return max_interval
+    return min(max_interval, threshold.total_seconds() * heuristic_interval_weight)
+
+
+async def handle_stale_sessions(
+    db: ExtendedAsyncSAEngine,
+    lifecycle_manager: SessionDestroyer,
+    status: SessionStatus,
+    threshold: TimeDelta,
+    interval: float = 0,  # NOTE: `aiotools.create_timer()` passes `interval` to its `cb`.
+) -> None:
+    now = datetime.now(tz=tzutc())
+    query = (
+        sa.select(SessionRow)
+        .where(SessionRow.status == status)
+        .where(
+            now
+            - SessionRow.status_history[status.name].astext.cast(sa.types.DateTime(timezone=True))
```

Comment:
make method in `SessionRow`

Reply:
https://github.com/lablup/backend.ai/pull/3992/commits/ea420f8f53a3c40d047ca00c02e4c4a288a78a02#diff-065d1c63c8c82f4548e2285001b8e4d29f17387ecc4bf80849816de6f3b07d4aR1314-R1316